### PR TITLE
Chunked txt fix

### PR DESCRIPTION
**Pull request by VSSRS**
I've drafted the incorporation of the ares_parse_txt_reply_ext API in Pycares and included some tests for it.
I encourage you to make a pass over the code, as I'm definitely not a C programmer.

Cheers.

### CODE REVIEW

**Comment by VSSRS** [comment]
Wait up. I've used a bad test case. According to https://tools.ietf.org/html/rfc4408#section-3.1.3 multiple strings should be concatenated WITHOUT a space. The SPF record of museumofplay.org is actually simply wrong (no spaces will be inserted, thus it is just not a valid SPF record. You can see that here: http://mxtoolbox.com/SuperTool.aspx?action=txt%3amuseumofplay.org&run=toolpage).

I'll change this in the code and try to find a more sensible test domain. 
There also seem to be some compilation issues with the code according to Appveyor (it compiled just fine on my Mac).

File: pycares/_cfficore/__init__.py [hunk 1]
```diff
@@ -279,23 +279,34 @@ def _query_cb(arg, status, timeouts, abuf, alen):
                 status = None
 
         elif query_type == _lib.T_TXT:
-            txt_reply = _ffi.new("struct ares_txt_reply **")
-            parse_status = _lib.ares_parse_txt_reply(abuf, alen, txt_reply);
+            tmp_obj = None
```
Comment (saghul) [inline pycares/_cfficore/__init__.py]: unused variable

File: pycares/_cfficore/__init__.py [hunk 2]
```diff
@@ -279,23 +279,34 @@ def _query_cb(arg, status, timeouts, abuf, alen):
...
+            assembled_text = ""
```
Comment (saghul) [inline pycares/_cfficore/__init__.py]: ditto

File: pycares/_cfficore/__init__.py [hunk 3]
```diff
@@ -651,7 +662,7 @@ def __init__(self, srv):
 
 class ares_query_txt_result(object):
     def __init__(self, txt):
-        self.txt = _ffi_string(txt.txt)
+        self.text = _ffi_string(txt.txt)
```
Comment (saghul) [inline pycares/_cfficore/__init__.py]: is this how the field is called in the C version?

File: src/cares.c [hunk 4]
```diff
@@ -39,7 +39,6 @@ struct sockaddr_in6 uv_ip6_addr(const char* ip, int port) {
     return addr;
 }
 
-
```
Comment (saghul) [inline src/cares.c]: please leave the style as is

File: src/cares.c [hunk 5]
```diff
@@ -486,16 +490,37 @@ query_txt_cb(void *arg, int status,int timeouts, unsigned char *answer_buf, int
         goto callback;
     }
 
+    first_record = 1;
+    assembled_txt = strdup("");
```
Comment (saghul) [inline src/cares.c]: don't use functions that allocate the memory by themselves, we want to use the Python mamory allocator. You can create a string with https://docs.python.org/2/c-api/string.html#c.PyString_FromString and then resize it with https://docs.python.org/2/c-api/string.html#c._PyString_Resize

File: src/cares.c [hunk 6]
```diff
@@ -486,16 +490,37 @@ query_txt_cb(void *arg, int status,int timeouts, unsigned char *answer_buf, int
         goto callback;
     }
 
+    first_record = 1;
```
Comment (saghul) [inline src/cares.c]: You should be able to not use a variable for this. Store a pointer to the current object (you already have a tmp_obj). If you get a record start and it's not NULL, append it to the list and create a new one. Same applies to the cffi changes.

File: tests/tests.py [hunk 7]
```diff
@@ -11,7 +11,7 @@
 class DNSTest(unittest.TestCase):
 
     def setUp(self):
-        self.channel = pycares.Channel(timeout=1.0, tries=1)
+        self.channel = pycares.Channel(timeout=5.0, tries=1)
```
Comment (saghul) [inline tests/tests.py]: why is this needed?

**Comment by saghul** [comment]
I did a first round of review, see the comments. Thanks for working on this one!

File: pycares/_cfficore/__init__.py (hunk 3 above)
Comment (VSSRS) [inline pycares/_cfficore/__init__.py]: https://github.com/saghul/pycares/blob/48fb58fac00079e0858987310ff8c0ad637b2c9f/src/pycares.h#L229

Comment (saghul) [inline pycares/_cfficore/__init__.py]: 👍

File: tests/tests.py (hunk 7 above)
Comment (VSSRS) [inline tests/tests.py]: Sometimes the tests on Travis CI failed (inconsistently) on the domains I used as examples for fragmented TXT records. I believe a timeout of 1 second was okay for google.com (used in the tests previously), but not necessarily for "less provisioned domains" that might not always respond within that timespan.

File: pycares/_cfficore/__init__.py (hunk 1 above)
Comment (VSSRS) [inline pycares/_cfficore/__init__.py]: Yes. Thanks.

File: pycares/_cfficore/__init__.py (hunk 2 above)
Comment (VSSRS) [inline pycares/_cfficore/__init__.py]: Yup.

File: tests/tests.py (hunk 7 above)
Comment (saghul) [inline tests/tests.py]: I see. Any chance we can test against some highly available server like Google? I know they are not going away anytime soon ;-)

File: pycares/_cfficore/__init__.py (hunk 3 above)
Comment (VSSRS) [inline pycares/_cfficore/__init__.py]: I'm guessing this will break code for people using CFFI? As they would've been using "txt" instead of "text" property on their results in the previous versions. Is that a problem?

Comment (saghul) [inline pycares/_cfficore/__init__.py]: No, I consider that a bug in the CFFI implementation which I didn't catch.

File: tests/tests.py (hunk 7 above)
Comment (VSSRS) [inline tests/tests.py]: I understand why you want to test this on a highly available server. However, since we're testing the actual contents of the TXT records in these tests (as opposed to tests that mainly just check the return type), we will very likely face the situation were the records will change at some point.
Finding a test for chunked TXT records on a server like Google, will not really help in that regard, as they are susceptible to change as well. Not sure which road to go with these tests. Any thoughts?

**Comment by VSSRS** [comment]
I tried to incorporate your comments. Feel free to take a look.

AppVeyor is refusing to build it at this time. Seems like a problem on their side since the branch has no conflicts according to GitHub.

File: src/cares.c [hunk 8]
```diff
@@ -486,16 +489,30 @@ query_txt_cb(void *arg, int status,int timeouts, unsigned char *answer_buf, int
         goto callback;
     }
 
+    tmp_obj = PyStructSequence_New(&AresQueryTXTResultType);
```
Comment (saghul) [inline src/cares.c]: you can set these to null and create them inside the loop as in the CFFI variant. Also: error checking.

File: pycares/_cfficore/__init__.py [hunk 9]
```diff
@@ -279,23 +279,32 @@ def _query_cb(arg, status, timeouts, abuf, alen):
...
             if parse_status != ARES_SUCCESS:
                 result = None
                 status = parse_status
             else:
                 result = []
-                txt_reply_ptr = _ffi.new("struct ares_txt_reply **")
+                txt_reply_ptr = _ffi.new("struct ares_txt_ext **")
                 txt_reply_ptr[0] = txt_reply[0]
+                tmp_obj = None
+                # first_record = True
```
Comment (saghul) [inline pycares/_cfficore/__init__.py]: remove uncommented line

File: pycares/_cfficore/__init__.py [hunk 10]
```diff
@@ -279,23 +279,32 @@ def _query_cb(arg, status, timeouts, abuf, alen):
...
                         break
-                    result.append(ares_query_txt_result(txt_reply_ptr[0]))
+                    if txt_reply_ptr[0].record_start == 1:
+                        if tmp_obj is not None: 
+                            result.append(tmp_obj)
+                        tmp_obj = ares_query_txt_result(txt_reply_ptr[0])
+                    else:
+                        new_chunk = ares_query_txt_result(txt_reply_ptr[0])
+                        tmp_obj.text += new_chunk.text
+                    # first_record = False
```
Comment (saghul) [inline pycares/_cfficore/__init__.py]: ditto

File: src/cares.c [hunk 11]
```diff
@@ -454,8 +455,10 @@ query_txt_cb(void *arg, int status,int timeouts, unsigned char *answer_buf, int
 {
     PyGILState_STATE gstate = PyGILState_Ensure();
     int parse_status;
-    struct ares_txt_reply *txt_reply, *txt_ptr;
-    PyObject *dns_result, *errorno, *tmp, *result, *callback;
+    struct ares_txt_ext *txt_reply, *txt_ptr;
+    PyObject *dns_result, *errorno, *tmp_obj, *result, *callback;
+    long assembled_ttl;
```
Comment (saghul) [inline src/cares.c]: no need to keep the assembled TTL, just set it when you create the structsequence since it's the same for all chunks

File: src/cares.c [hunk 12]
```diff
@@ -486,16 +489,34 @@ query_txt_cb(void *arg, int status,int timeouts, unsigned char *answer_buf, int
         goto callback;
     }
 
+    tmp_obj = NULL;
+    assembled_txt = NULL;
     for (txt_ptr = txt_reply; txt_ptr != NULL; txt_ptr = txt_ptr->next) {
-        tmp = PyStructSequence_New(&AresQueryTXTResultType);
-        if (tmp == NULL) {
+        // Initialzing new assembled record if necessary
```
Comment (saghul) [inline src/cares.c]: please use /* */ for comments, like the rest of the code

File: src/cares.c [hunk 13]
```diff
@@ -486,16 +489,34 @@ query_txt_cb(void *arg, int status,int timeouts, unsigned char *answer_buf, int
...
+        PyBytes_ConcatAndDel(&assembled_txt, PyBytes_FromString((char*)txt_ptr->txt));
+        if (assembled_txt == NULL) {
             break;
         }
-        PyStructSequence_SET_ITEM(tmp, 0, Py_BuildValue("s", (const char *)txt_ptr->txt));
-        PyStructSequence_SET_ITEM(tmp, 1, PyInt_FromLong((long)txt_ptr->ttl));
-        PyList_Append(dns_result, tmp);
-        Py_DECREF(tmp);
+        // Add previously assembled record to dns_result if necessary
+        if (txt_ptr->next == NULL || txt_ptr->next->record_start == 1) {
```
Comment (saghul) [inline src/cares.c]: peaking into the next element here is not nice. Maybe it's jyst better to rewrite the for loop as for(;;) ... like the CFFI version.

File: src/cares.c [hunk 14]
```diff
@@ -486,16 +489,34 @@ query_txt_cb(void *arg, int status,int timeouts, unsigned char *answer_buf, int
...
+            PyStructSequence_SET_ITEM(tmp_obj, 0, Py_BuildValue("s", PyBytes_AsString(assembled_txt)));
```
Comment (saghul) [inline src/cares.c]: You already have a bytes object, you should be able to just set it on the sequence without getting its internal string, which will end up making a new copy of it.

File: tests/tests.py (hunk 7 above)
Comment (saghul) [inline tests/tests.py]: Once we know it works I'll probably comment them out. Don't worry about that.

**Comment by saghul** [comment]
Left some more comments, we're close!

File: src/cares.c (hunk 14 above)
Comment (VSSRS) [inline src/cares.c]: I did this originally, but it created inconsistencies between Python 2 and 3. Without the Py_BuildValue, the result will be in 'bytes' in Python3 and Python2 would just be 'str'. What behaviour would be most desirable here?

Comment (saghul) [inline src/cares.c]: Ah true that. Then maybe use AS_STRING variant, which is cheaper because it doesn't do any checks. But we know the object is ok.

File: pycares/_cfficore/__init__.py [hunk 15]
```diff
@@ -279,23 +279,30 @@ def _query_cb(arg, status, timeouts, abuf, alen):
...
                 status = parse_status
             else:
                 result = []
-                txt_reply_ptr = _ffi.new("struct ares_txt_reply **")
+                txt_reply_ptr = _ffi.new("struct ares_txt_ext **")
                 txt_reply_ptr[0] = txt_reply[0]
+                tmp_obj = None
                 while True:
                     if txt_reply_ptr[0] == _ffi.NULL:
+                        result.append(tmp_obj)
```
Comment (saghul) [inline pycares/_cfficore/__init__.py]: let's go defensive here and check if tmp_obj is not None, just in case.

File: src/cares.c [hunk 16]
```diff
@@ -486,16 +488,40 @@ query_txt_cb(void *arg, int status,int timeouts, unsigned char *answer_buf, int
...
     }
 
-    for (txt_ptr = txt_reply; txt_ptr != NULL; txt_ptr = txt_ptr->next) {
-        tmp = PyStructSequence_New(&AresQueryTXTResultType);
-        if (tmp == NULL) {
+    tmp_obj = NULL;
+    assembled_txt = NULL;
+    txt_ptr = txt_reply;
+    while (1) {
+        if (txt_ptr == NULL || (txt_ptr->record_start == 1 && tmp_obj != NULL)) {
```
Comment (saghul) [inline src/cares.c]: it's possible you end up here having txt_ptr and tmp_obj both NULL.

File: src/cares.c [hunk 17]
```diff
@@ -486,16 +488,40 @@ query_txt_cb(void *arg, int status,int timeouts, unsigned char *answer_buf, int
...
+                break;
+            }
+            /* ttl of the first chunk is representative for the entire record */
+            PyStructSequence_SET_ITEM(tmp_obj, 1, PyInt_FromLong((long)txt_ptr->ttl)); 
+            assembled_txt = PyBytes_FromString("");
+        }
+        /* Concatenate each chunk's text onto the assembled record */
+        PyBytes_ConcatAndDel(&assembled_txt, PyBytes_FromString((char*)txt_ptr->txt));
+        if (assembled_txt == NULL) {
             break;
```
Comment (saghul) [inline src/cares.c]: in this case you want to decref the struct sequence, or you'd leak it

**Comment by saghul** [comment]
Great work @VSSRS, this is ready now! I'll merge it later tonight. If you can squash all your commits into one that would be great, otherwise I'll do it during merge.

**Comment by VSSRS** [comment]
My pleasure :-) Thanks for the quick feedback.

**Comment by saghul** [comment]
👍

**Comment by saghul** [comment]
Landed in :sparkles: 65417ef :sparkles: Thanks for your patience and work!
